### PR TITLE
Update external codelists

### PR DESCRIPTION
**Pull request by jpmckinney**

### CODE REVIEW

**Comment by jpmckinney** [comment]
For dosage form, these are the common codes:

| original code | original title | new code | new title |
| - | - | - | - |
| cream | Cream | CRM | Cream |
| drops | Drops | DROP | Drops |
| gas | Gas | GASINHL | Gas for Inhalation |
| gel | Gel | GEL | Gel |
| granules | Granules | GRAN | Granules |
| foam | Foam | FOAM | Foam |
| lotion | Lotion | LTN | Lotion |
| ointment | Ointment | OINT | Ointment |
| paste | Paste | PASTE | Paste |
| pillules | Pillules | PILL | Pill |
| powder | Powder | POWD | Powder |
| solution | Solution | SOL | Solution |
| spray | Spray | SPRY | Sprays |
| suppository | Suppository | SUPP | Suppository |
| suspension | Suspension | SUSP | Suspension |

These are present in HL7, but at a different level of specificity, and are therefore excluded:

| original code | original title | reason |
| - | - | - |
| capsule | Capsule | child of Pill |
| tablet | Tablet | child of Pill |
| syrup | Syrup | child of Oral Solution and grandchild of Solution |
| inhaler | Inhaler | child of Inhalant (gas/solid) or Aerosol (gas/liquid) |
| liquid | Liquid | non-selectable parent of Liquid Cleanser, Oil, Solution |

These are excluded from the new codelist:

| original code | original title | notes |
| - | - | - |
| poultice | Poultice | |
| insert | Insert | covered by Suppository? |
| pessary | Pessary | covered by Suppository |
| implant | Implant | an administration route, not a dosage form? |
| injection | Injection | an administration route, not a dosage form |

These sprays are not children of SPRY (Sprays), for whatever reason. They are excluded from the new codelist.

| new code | new title | notes |
| - | - | - |
| PWDSPRY | Powder Spray | |
| SPRYADAPT | Spray with Adaptor | unclear why adaptor is relevant |
| DERMSPRY | Dermal Spray | mixes concept of administration route |
| RECSPRY | Rectal Spray | mixes concept of administration route |
| VAGSPRY | Vaginal Spray | mixes concept of administration route |

These were not included in the original codelist:

| new code | new title |
| - | - |
| AER | Aerosol |
| APPFUL | Applicatorful |
| BAR | Bar |
| BEAD | Beads |
| CAKE | Cake |
| CEMENT | Cement |
| CRYS | Crystals |
| DISK | Disk |
| FLAKE | Flakes |
| GUM | ChewingGum |
| INHL | Inhalant |
| LIQCLN | Liquid Cleanser |
| OIL | Oil |
| PAD | Pad |
| PATCH | Patch |
| PELLET | Pellet |
| PUFF | Puff |
| SCOOP | Scoops |
| SWAB | Swab |
| WAFER | Wafer |

**Comment by jpmckinney** [comment]
For container, preFilledSyringe (Pre-Filled Syringe) is not included, and the following are added:

| new code | new title |
| - | - |
| box | Box |
| can | Can |
| disk | Disk |
| doset | Dosette |
| jug | Jug |
| minim | Minim |
| nebamp | Nebule Amp |
| ovul | Ovule |
| pch | Pouch |
| pkt | Packet |
| tub | Tub |

**Comment by jpmckinney** [comment]
HL7's administration route is quite different (2x as many codes).

**Comment by jpmckinney** [comment]
@yolile Outside this PR I also made this small commit: b410711f82da21e0104ff3938d57584c8667c461

**Comment by jpmckinney** [comment]
Hmm, I've now discovered https://standardterms.edqm.eu (I have a free account) via [this presentation](https://www.dgra.de/media/pdf/fortbildung/kongresse/2016/kongress2016-09-Franken.pdf), which also covers administration route, container, and dosage form. I'll do a comparison...

**Comment by jpmckinney** [comment]
@yolile I'm beginning to wonder whether we need to have identifier objects for these codelists. I think it is not practical or reasonable for countries to map codelists with hundreds of codes to our lists. What do you think?

**Update:** Alternately, it is free text (a simple string), and the publisher is recommended to name the codelists used, and provide access to tables where the codes can be looked up.

File: manage.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+def update_dosage_form():
+    """
+    Update schema/codelists/dosageForm.csv from HL7.
+    """
+    # https://terminology.hl7.org/CodeSystem/v3-orderableDrugForm/
+    codes, not_selectable = hl7('orderableDrugForm')
+
+    with csv_dump('dosageForm.csv', ['Code', 'Title', 'Description']) as writer:
+        for code in codes:
+            if 'SPRY' in code['code'] or code['code'] != 'SPRY':
```
Comment (yolile) [inline manage.py]: `or code['code'] != 'SPRY'` is this correct?

Comment (jpmckinney) [reply manage.py]: Fixed in 35924ef

**Comment by yolile** [comment]
> @yolile I'm beginning to wonder whether we need to have identifier objects for these codelists. I think it is not practical or reasonable for countries to map codelists with hundreds of codes to our lists. What do you think?

As we are using HL7 everywhere now, I think that still makes sense to recommend to use codes, as HL7 is for ensuring interoperability, and for doing that the preferred option is to use common codes

**Comment by jpmckinney** [comment]
> > @yolile I'm beginning to wonder whether we need to have identifier objects for these codelists. I think it is not practical or reasonable for countries to map codelists with hundreds of codes to our lists. What do you think?
> 
> As we are using HL7 everywhere now, I think that still makes sense to recommend to use codes, as HL7 is for ensuring interoperability, and for doing that the preferred option is to use common codes

Okay, maybe we just add a paragraph to the readme to say that if you haven't adopted / can't map to HL7, then you can use your own codes, and use the data guide/publication policy to name the codelists used, and provide access to tables where the codes can be looked up.

**Update:** We can also add a paragraph to explain why standardized codes is important. From your review document:

> The ideal unit for comparing prices is a quality-constant course of therapy. In fact, the range and mix of dosage forms, strengths and pack sizes differ significantly across countries.  Most existing studies compare the price for a single pack for each product, usually selecting a typical pack in the base country.  But this pack may be atypical or totally unavailable in other countries; even if it is available, the distribution of pack sizes and their relative prices may differ, leading to differences in the weighted average price. 
(17) (PDF) International Price Comparisons for Pharmaceuticals. Available from: https://www.researchgate.net/publication/257901972_International_Price_Comparisons_for_Pharmaceuticals

File: manage.py (hunk 1 above)
Comment (yolile) [reply manage.py]: so we only want the codes that contain 'SPRY'?

**Review by yolile — APPROVED** [review]

**Comment by jpmckinney** [comment]
Looking at https://standardterms.edqm.eu, the following codelists don't seem to be relevant.

[Pharmaceutical Dose Forms by State Of Matter](https://standardterms.edqm.eu/browse/get_concepts_by/SOM/SOM) and [by Intended Site](https://standardterms.edqm.eu/browse/get_concepts_by/ISI/PDF) mix dosage form with administration route (Concentrate for oral solution) and administration method (Concentrate for gargle), so they don't follow our desired semantics.

[Unit of presentation](https://standardterms.edqm.eu/browse/get_concepts/UOP) has terms like Actuation, which suggest that the codelist is about how the medicine is presented or administered to the patient, not how it's acquired or ordered. It overlaps with EDQM's container codelist (see next comment).

> A unit of presentation used to represent the quantity of product that is administered following a single operation of a metered-dose pump, valve or other equivalent dosing mechanism.
> For example, to describe the strength of a pressurised inhalation solution (e.g. via an inhaler): "contains 100 micrograms per actuation".

Some other irrelevant lists:

* Combined Pharmaceutical Dose Forms: List of combined pharmaceutical dose forms, each of which represents a combination of two or more pharmaceutical dose forms that are intended to be reconstituted together to form a single administrable pharmaceutical dose form.
  *  Examples: "Concentrate and solvent for cutaneous solution"
* Combined Terms: List of combined terms, each of which is a combination of one or more pharmaceutical dose forms or combined pharmaceutical dose forms with one or more items of packaging.
  * Examples: "Bee-hive dispersion in sachet"
* Combination Packs: List of combination packs, each of which represents a combination of two or more products that are packaged together but that are intended to be administered independently of each other.
  *  Examples: "Capsule, hard + tablet", "Cream + pessary"
* Patient-Friendly Terms by Intended Site: List of patient-friendly terms, arranged according to their intended site of administration. These are for use only where authorised, and where labelling space on the immediate packaging is limited such that the full Standard Term cannot be used.
* Mapped Terms: List of mapped terms, arranged according to their source. These are not Standard Terms, but terms from external databases that are referenced here in order to help users identify potential equivalent Standard Terms. The EDQM is not responsible for the accuracy or completeness of any mapping. Questions about any terms should be directed to the owners of the external database.

**Comment by jpmckinney** [comment]
Looking at Packaging at https://standardterms.edqm.eu:

[Packaging](https://standardterms.edqm.eu/browse/get_concepts_by/PAC/PAC) is divided into Administration device (Nozzle), Closure (Child-resistant closure) and Container. For containers for human use (not veterinary only), the same words are used for:

| HL7 | EDQM |
| - | - |
| Ampule | Ampoule |
| Bag | Bag |
| Blister Pack | Blister |
| Bottle | Bottle |
| Box | Box |
| Cartridge | Cartridge |
| Jar | Jar |
| Pouch | Pouch |
| Sashet | Sachet |
| Strip | Strip |
| Tube | Tube |
| Vial | Vial |

Codes in HL7 but not EQDM:

| HL7 | notes |
| - | - |
| Nebule Amp | a specific type of Ampoule for a nebulizer |
| Minim | seems to be an Ampoule/Vial with a specific volume |
| Dosette | might be the same as Tablet container in EQDM |
| Disk | might be covered by Multidose container in EQDM https://99percentinvisible.org/episode/repackaging-the-pill/ |
| Packet | unclear how it differs from Sachet: "Container consisting of two surfaces made of flexible material to be closed only by sealing or folding over. The contents are intended for single use." |
| Jug | unclear how it differs from Bottle: "Container with a more or less pronounced neck and usually a flat bottom." |
| Tub | unclear how it differs from Jar: "Container, without a pronounced neck, with a wide opening at the top and a more-or-less flat bottom. It is suited for semi-solid and solid pharmaceutical forms (including those that are stored/supplied in a liquid medium, such as living tissue equivalents). It can be reclosed." |
| Tin | unclear how it differs from Bottle, besides the material being tin |
| Can | unclear how it differs from Tub, except maybe it can't be reclosed |
| Ovule | not sure what it is |

26 codes in EDQM but not HL7:

<details>

| Term | Definition |
| - | - |
| Container | An item of packaging that is used for the storage, identification and/or transport of a medicinal product. Only to be used where more-specific terms do not apply. Specific characteristics of the container are defined in the product information of the medicinal product." |
| Unit-dose blister | Container consisting of two layers of which one is shaped to contain the individual unit, usually provided with perforations, the intention being to allow each individual unit to be separated for single unit administration. All the information required for blister packs must appear on each unit dose presentation. Blisters and strips are excluded." |
| Single-dose container | A container that holds a quantity of the preparation intended for total or partial use on one occasion only. |
| Barrel | Container of large size, suited for liquid, solid and semi-solid pharmaceutical forms." |
| Dose-dispenser cartridge | Container intended for use in a dose dispenser, usually pre-filled with a medicinal product. Cartridges, intended for use in pens, are excluded." |
| Dredging container | Container for a pharmaceutical form to be applied by dredging. |
| Dropper container | A container, usually a bottle, fitted with a dropper applicator." |
| Gas cylinder | Container usually cylindrical suited for compressed, liquefied or dissolved gas, fitted with a device to regulate the spontaneous outflow of gas at atmospheric pressure and room temperature." |
| Pressurised container | Container suited for compressed, liquefied or dissolved gas fitted with a device to enable, after its actuation, a controlled spontaneous release of the contents at atmospheric pressure and room temperature." |
| Fixed cryogenic vessel | A static thermally insulated container designed to maintain the contents in the liquid state |
| Mobile cryogenic vessel | A mobile thermally insulated container designed to maintain the contents in a liquid state. |
| Multidose container with airless pump | Multidose container with an integral pump designed to protect the contents against in use-contamination. |
| Multidose container with metering pump | Multidose container with integral metering pump. |
| Multidose container with pump | Multidose container with an integral pump. Metering pump, spray pump, and multidose container with airless pump are excluded." |
| Pre-filled gastroenteral tube | Pre-filled tube for the administration of a medicinal product to the gastroenteral tract. |
| Pre-filled injector | Filled container, usually fitted with an injection needle, which may or may not be fitted with a needle safety device, containing a sterile, single-dose parenteral preparation. This term is only to be used when a more-specific term such as 'Pre-filled syringe' or 'Pre-filled pen' is not appropriate." |
| Pre-filled oral applicator | Filled container for administering a liquid or semi-solid pharmaceutical form to the oral cavity, containing the final preparation in one compartment, or the components necessary for its preparation in different compartments. This term may be used to describe a pre-filled oral syringe in cases where there is a risk of administration error, e.g. where a similar product is often administered by injection, such as with vaccines." |
| Pre-filled oral syringe | Filled container for administering a liquid or semi-solid pharmaceutical form to the oral cavity, containing the final preparation in one compartment, or the components necessary for its preparation in different compartments." |
| Pre-filled pen | Filled container consisting usually of a cartridge, usually fitted with an injection needle, housed in a pen device that may or may not be fitted with a needle safety device. The cartridge contains a sterile, single-dose or multidose, parenteral preparation; it contains the final preparation in one compartment, or the components necessary for its preparation in different compartments. The cartridge is not replaceable and the pre-filled pen is intended to be discarded when empty. The pre-defined dose, which may or may not be adjustable, is administered via an actuation, such as the pressing of a button." |
| Pre-filled syringe | Filled container consisting of an injection syringe, generally supplied with an injection needle, which may or may not be fitted with a needle safety device. It contains a sterile, single-dose or multidose, parenteral preparation; it contains the final preparation in one compartment, or the components necessary for its preparation in different compartments. The pre-filled syringe is intended to be discarded when empty." |
| Roll-on container | A container, usually a bottle, fitted with a roll-on applicator" |
| Spray container | Container for a liquid pharmaceutical form to be converted into a spray by mechanical means. |
| Tablet container | Container without neck and with a flat bottom, suited for tablets, capsules, etc., can be re-closed well." |
| Straw | Hollow tube containing a single dose of a medicinal product for oral administration by sucking. |
| Wrapper | Single-dose or multidose container, usually consisting of a single flexible surface (which may consist of one or more layers and may be coated) intended to be used to enclose one or more units of a solid preparation such as a tablet or lozenge. Multidose wrappers may be used to present the individual units of a medicinal product together in the form of a roll that can be torn or otherwise opened to allow the removal of the required number of units. Blisters, sachets, strips and unit-dose blisters are excluded." |

</details>

**Comment by jpmckinney** [comment]
Here is the EDQM route codelist (81 codes). At first glance, EDQM's list seems to make a better distinction between administration route (anatomy) and administration method (process), which HL7 does not.

I haven't compared whether HL7 and EDQM cover the same anatomical routes.

That said, EDQM names the list "Routes and Methods of Administration", but it doesn't have common methods like injection or insertion. So, the list seems incomplete. The only routes/methods not of the form "[anatomy] use" are:

* Haemodialysis
* Implantation
* Infiltration
* Iontophoresis
* Route of administration not applicable
* Skin scarification

The (["officially phased out"](https://www.fda.gov/drugs/electronic-regulatory-submission-and-review/data-standards-manual-monographs)) FDA list looks similar to EDQM's, at first glance: https://www.fda.gov/drugs/data-standards-manual-monographs/route-administration

<details>

| Term | Definition |
| - | - |
| Auricular use | Administration of a medicinal product to the ear. |
| Buccal use | Administration of a medicinal product to the buccal cavity (pouch located between the cheek and the gum) to obtain a systemic effect |
| Cutaneous use | Administration of a medicinal product to the skin and/or cutaneous wounds and/or nails and/or hair in order to obtain a local effect. |
| Dental use | "Administration of a medicinal product to and/or in the teeth or, on and/or around the nerves supplying the teeth." |
| Endocervical use | Administration of a medicinal product to the cervix uteri. |
| Endosinusial use | Administration of a medicinal product to the sinuses to obtain a local effect. |
| Endotracheopulmonary use | Administration of a medicinal product to the trachea and/or bronchi by instillation (preparations for inhalation are excluded; see inhalation use). |
| Epidural use | Injection of a medicinal product into the epidural space. |
| Epilesional use | Administration of a medicinal product onto a lesion. |
| Extraamniotic use | Injection of a medicinal product between chorion and amnion. |
| Extracorporeal use | Use of a medicinal product outside the body |
| Gastric use | Administration of a medicinal product to the stomach by means of an appropriate device |
| Gastroenteral use | Administration of a medicinal product to the stomach or duodenum gastroenteral tract by means of an appropriate device. For use only when gastric use and intestinal use do not apply. |
| Gingival use | Administration of a medicinal product to the gingivae. |
| Haemodialysis | Clearance of the blood by means of a semipermeable membrane. [Previous English term 'Hemodialysis' corrected 04/07/11.] |
| Implantation | Insertion of an implant or living tissue equivalent into living tissue. |
| Infiltration | "Method of administration, usually by injection, whereby a fluid passes into a target tissue (e.g. anaesthetic infiltration)" |
| Inhalation use | Administration of a medicinal product to the respiratory system by inhalation to obtain a systemic or a local effect in the lower respiratory tract. Nasal use and endotracheopulmonary use are excluded |
| Intestinal use | "Administration of a medicinal product to the intestine (duodenum, jejunum, ileum and colon) by means of an appropriate device. Gastroenteral use is excluded" |
| Intraamniotic use | Injection of a medicinal product into the amniotic cavity. |
| Intraarterial use | Injection of a medicinal product into an artery. |
| Intraarticular use | Injection of a medicinal product into an articular cavity. |
| Intrabursal use | Injection of a medicinal product into bursae and tendons. |
| Intracameral use | Administration of a medicinal product directly into the anterior chamber of the eye |
| Intracardiac use | Injection of a medicinal product into the cardiac muscle and/or cardiac cavity. |
| Intracartilaginous use | Administration of a medicinal product into the cartilage |
| Intracavernous use | Injection of a medicinal product into the corpus cavernosum. |
| Intracerebral use | Administration of a medicinal product directly to the brain tissue. This term is only for use when a more-specific term (e.g. 'Intraputaminal use') is not applicable. |
| Intracerebroventricular use | Administration of a medicinal product into the cerebral ventricles (cerebral ventricular system) of the brain. |
| Intracervical use | Injection of a medicinal product into the cervix uteri. |
| Intracholangiopancreatic use | "Injection of a medicinal product into the bile duct and the pancreatic duct, for example via a cannula introduced into the ampulla of Vater (the common opening of the ducts), usually for the administration of a contrast medium for techniques such as endoscopic retrograde cholangiopancreatography (ERCP)." |
| Intracisternal use | Administration of a medicinal product into the cisterna magna. |
| Intracorneal use | Administration of a medicinal product into the cornea of the eye. |
| Intracoronary use | Injection of a medicinal product into the coronary artery. |
| Intradermal use | Injection of a medicinal product into the dermis. |
| Intradiscal use | Injection of a medicinal product into the nucleus pulposus of an intervertebral disc. |
| Intraepidermal use | Administration of a medicinal product into the epidermis |
| Intraglandular use | "Administration of a medicinal product directly into a gland, usually by injection. Only to be used where more-specific terms such as 'Intraprostatic use' and 'Intramammary use' do not apply." |
| Intralesional use | Administration by injection or any other means of a medicinal product directly to a lesion. |
| Intralymphatic use | Injection of a medicinal product into a lymphatic vessel. |
| Intramuscular use | Injection of a medicinal product into muscular tissue. |
| Intraocular use | "Administration of a medicinal product into the eye. The term 'intraocular use' is only for use when a more specific term (e.g. 'intracameral use', 'intravitreal use') does not apply. Ocular use and subconjunctival use are excluded." |
| Intraosseous use | Administration of a medicinal product into the bone marrow. Intrasternal use is excluded. |
| Intrapericardial use | Injection of a medicinal product to the pericardium. |
| Intraperitoneal use | Injection of a medicinal product into the peritoneal cavity. |
| Intrapleural use | Injection of a medicinal product into the pleural cavity. |
| Intraportal use | Injection/infusion of a medicinal product into the hepatic portal vein for local delivery to the liver. |
| Intraprostatic use | Administration of a medicinal product into the prostate |
| Intraputaminal use | Administration of a medicinal product into one or both of the putamen of the brain. |
| Intrasternal use | Injection of a medicinal product into the bone marrow of the sternum. |
| Intrathecal use | Injection of a medicinal product through the dura to the subarachnoid cavity. |
| Intratumoral use | Injection of a medicinal product into a tumor. |
| Intrauterine use | Administration of a medicinal product to the cavity of the uterus. |
| Intravenous use | Injection of a medicinal product into a vein. |
| Intravesical use | Administration of a medicinal product to the urinary bladder. |
| Intravitreal use | Administration of a medicinal product into the rear chamber of the eye. |
| Iontophoresis | Introduction of (an) ionised active substance(s) through the intact skin by application of a direct electric current |
| Laryngopharyngeal use | Administration of a medicinal product to the laryngopharynx for a local effect (anaesthetics). |
| Nasal use | Administration of a medicinal product to the nose to obtain a systemic or local effect. Inhalation therapy intended for the lower respiratory tract is excluded; see inhalation use. |
| Ocular use | Administration of a medicinal product upon the eyeball and/or conjunctiva. |
| Oral use | Taking a medicinal product by means of swallowing. |
| Oromucosal use | "Administration of a medicinal product to the oral cavity to obtain either a systemic or a local effect. The term oromucosal is only for use when a more specific term (e.g. buccal, gingival, sublingual...) does not apply. Oral use is excluded." |
| Oropharyngeal use | Administration of a medicinal product to the oropharynx (throat) to obtain a local effect. |
| Periarticular use | Injection of a medicinal product around a joint. |
| Perineural use | Injection of a medicinal product into the direct surroundings of one or more nerves. |
| Periodontal use | Administration to the pouch between the tooth and the gingiva. |
| Periosseous use | Administration of a medicinal product on or around the bone |
| Peritumoral use | Administration of a medicinal product directly into the region surrounding a tumour. |
| Posterior juxtascleral use | "Administration of a medicinal product beneath the sub-tenon membrane of the sclera (i.e. in the episcleral space), adjacent to the macula." |
| Rectal use | Administration of a medicinal product to the rectum in order to obtain a local or systemic effect. |
| Retrobulbar use | Administration of a medicinal product behind the eyeball |
| Route of administration not applicable | "Applies to medicinal products not directly coming into contact with the body of the patient, or administration to various or non-specified anatomical sites." |
| Skin scarification | Administration of a medicinal product by scratching the skin. |
| Subconjunctival use | Injection of a medicinal product underneath the conjunctiva. |
| Subcutaneous use | "Injection of a medicinal product directly underneath the skin, i.e. subdermally" |
| Sublingual use | Administration of a medicinal product under the tongue to obtain a systemic effect. |
| Submucosal use | Injection of a medicinal product directly underneath a mucosa |
| Subretinal use | Administration of a medicinal product between the sensory retina (neural retina) and the retinal pigment epithelium of the eye. |
| Transdermal use | Administration of a medicinal product to the skin in order to obtain a systemic effect after passing through the skin barrier. |
| Urethral use | Administration of a medicinal product to the urethra. |
| Vaginal use | Administration of a medicinal product to the vagina. |

</details>

**Pull request merged**